libs/wingc/src/jsify.rs

```diff
@@ -41,21 +42,190 @@ const $App = __app(process.env.WING_TARGET);
+						match r {
+							Reference::Identifier(s) => {
+								if s.name == "this" {
+									// resolve the type of the resource by looking up "this" in our environment
+									let resource_type = self.env
+										.lookup(s, None).expect("'this' to be found")
+										.as_variable().expect("'this' to be a variable")
+										.type_;
+
+									// lookup our field in the class environment
+									match &*resource_type {
+            				Type::Resource(r) => {
+											let field_kind = r.env
+												.lookup(property, None).expect("unable to find field in resource env")
+												.as_variable().expect("field is not a variable");
+
+											// we only care about preflight fields (inflight fields are normal references)
+											if field_kind.flight == Phase::Preflight {
+
+												// don't allow capturing reassignable (`var`) fields.
+												if field_kind.reassignable {
+													self.diagnostics.insert(Diagnostic {
+														level: DiagnosticLevel::Error,
+														message: format!(
+															"Unable to reference \"this.{}\" from inflight method \"{}\" because it is reassignable (\"var\")",
+															property.name, 
+															self.method_name.name,
+														),
+														span: Some(property.span.clone()),
+													});
+													return;
+												}
+
+												// check if the type is capturable (resource, primitive or immutable collection)
+												if ! field_kind.type_.is_capturable() {
+													self.diagnostics.insert(Diagnostic {
+														level: DiagnosticLevel::Error,
+														message: format!(
+															"Unable to reference \"this.{}\" from inflight method \"{}\" because type {} is not capturable",
+															property.name, 
+															self.method_name.name,
+															field_kind.type_,
+														),
+														span: Some(property.span.clone()),
+													});
+
+													return;
+												}
+
+												debug!("field \"this.{}\" is referenced from inflight method {}", property.name, self.method_name.name);
+
+												// get/create an "ops" set for this field
+												let mut ops = if let Some(ops) = self.references.get(&property.name) {
+													ops.clone()
+												} else {
+													IndexSet::new()
+												};
+
+												if ! self.path.is_empty() {
+													let op = self.path[0].clone();
+
+													// if we are referencing a resource, verify that the operation is an inflight method
+													if let Some(r) = field_kind.type_.as_resource() {
+														let mut found_method = false;
+
+														for (name, typeref) in r.methods(true) {
+															if name == op {
+																match &*typeref {
+																	Type::Function(f) => {
+																		if f.flight == Phase::Inflight {
+																			found_method = true;
+																		}
+																	}
+																	_ => {}
+																}
```

Comment:
I think we can use `typeref.as_function_sig()` here

Reply:
done